### PR TITLE
refactor: use new ingester<>querier wire protocol

### PR DESCRIPTION
**Pull request by crepererum**
Use and document the new and more flexible ingester<>querier wire
protocol.

Note that the ingester does NOT stream the response data yet, but the
internal data structures would allow that. A follow-up change will
adjust the ingester code to stream the data.

Ref #4849.

### CODE REVIEW

File: docs/ingester_querier_protocol.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
+# Ingester ⇔ Querier Query Protocol
```
Comment (alamb) [inline docs/ingester_querier_protocol.md:1]: Thank you for this document

File: docs/ingester_querier_protocol.md [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    - Persistence Information:
+      - Sequence number of max. persisted parquet file
+      - Sequence number of max. persisted tombstone
+    - For each chunk **(B)**:
+      - Record batches with following operations applied **(C)**:
+        - selection (i.e. row filter via predicates)
+        - projection (i.e. column filter)
+        - tombstone materialization
+
+This is mapped to the following the following stream of Flight messages:
```
Comment (alamb) [inline docs/ingester_querier_protocol.md:35]:
```suggestion
This is mapped to the following stream of Flight messages:
```

File: docs/ingester_querier_protocol.md [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+
+## Example
+Imagine the following ingester state:
+
+- sequencer S1:
+  - partition P1:
+    - max. persisted parquet file at `sequence_number=10`
+    - max. persisted tombstone at `sequence_number=11`
+    - chunks C1 and C2
```
Comment (alamb) [inline docs/ingester_querier_protocol.md:61]: I think the ingester calls these `Snapshot`s not Chunks -- using the `chunk` term may confuse things

File: ingester/src/data.rs [hunk 4]
```diff
@@ -1530,53 +1533,126 @@ pub struct PartitionStatus {
...
 
-    /// Map each record batch to a partition ID.
-    pub batch_partition_ids: Vec<PartitionId>,
+/// Response streams for querier<>ingester requests.
+///
+/// The data structure is constructed to allow lazy/streaming data generation. For easier consumption according to the
+/// wire protocol, use the [`flatten`](Self::flatten) method.
+pub struct IngesterQueryResponse {
+    /// Stream of partitions.
+    pub partitions: Pin<Box<dyn Stream<Item = Result<IngesterQueryPartition, ArrowError>> + Send>>,
```
Comment (alamb) [inline ingester/src/data.rs:1566]: I wonder if using a typedef like 

```rust
type IngesterResponseStream = Pin<Box<dyn Stream<Item = Result<IngesterQueryPartition, ArrowError>> + Send>>;
```
Might make the code easier to read 🤔

Comment (alamb) [reply ingester/src/data.rs:1566]: What do you think about making this non `pub` -- as it seems like the `FlatIngesterQueryResponse` is the real public interface

File: ingester/src/data.rs [hunk 5]
```diff
@@ -1530,53 +1533,126 @@ pub struct PartitionStatus {
...
+        /// Partition ID.
+        partition_id: PartitionId,
+
+        /// Partition persistence status.
+        status: PartitionStatus,
+    },
+
+    /// Start a new chunk.
+    ///
+    /// The chunk is assigned to the partition of the last [`StartPartition`](Self::StartPartition) message.
```
Comment (alamb) [inline ingester/src/data.rs:1643]:
```suggestion
    /// The chunk belongs  to the partition of the last [`StartPartition`](Self::StartPartition) message.
```

File: ingester/src/querier_handler.rs [hunk 6]
```diff
@@ -144,45 +156,17 @@ pub async fn prepare_data_to_querier(
...
     );
-    debug!(?unpersisted_partitions);
     ensure!(
         !unpersisted_partitions.is_empty(),
         TableNotFoundSnafu {
             namespace_name: &request.namespace,
             table_name: &request.table
         },
     );
-    debug!(
```
Comment (alamb) [inline ingester/src/querier_handler.rs:155]: 👌  very nice

File: ingester/src/querier_handler.rs [hunk 7]
```diff
@@ -654,19 +624,25 @@ mod tests {
...
-        assert_eq!(batches.len(), partition_ids_unique.len());
+    async fn ingester_response_to_record_batches(
+        response: IngesterQueryResponse,
+    ) -> Vec<RecordBatch> {
+        response
+            .flatten()
+            .try_filter_map(|msg| async move {
+                match msg {
+                    FlatIngesterQueryResponse::RecordBatch { batch } => Ok(Some(batch)),
+                    FlatIngesterQueryResponse::StartChunk { .. }
```
Comment (alamb) [inline ingester/src/querier_handler.rs:640]: I wonder if it is worth checking here that the schema in the batches matches what was sent in the previous `StartChunk` message?

File: querier/src/ingester/flight_client.rs [hunk 8]
```diff
@@ -145,206 +128,12 @@ where
     }
 }
 
-/// Protocol state.
```
Comment (alamb) [inline querier/src/ingester/flight_client.rs:148]: 🔥

**Review by alamb — APPROVED** [review]
This is kind of neat -- do I read this as the ingester code creates a `IngesterQueryResponse` -- and the querier uses a `IngesterQueryResponse` and that the grpc / futures plumbing arranges to send it over the wire?  That is very cool because then the `query_tests` still use the same streaming interface (even though it doesn't use the serialization part)

**Comment by alamb** [comment]
Really clever @crepererum

File: ingester/src/data.rs (hunk 4 above)
Comment (crepererum) [reply ingester/src/data.rs]: fixed (both)

File: docs/ingester_querier_protocol.md (hunk 3 above)
Comment (crepererum) [reply docs/ingester_querier_protocol.md]: fixed

File: ingester/src/querier_handler.rs (hunk 7 above)
Comment (crepererum) [reply ingester/src/querier_handler.rs]: done

**Comment by crepererum** [comment]
> do I read this as the ingester code creates a `IngesterQueryResponse` -- and the querier uses a `IngesterQueryResponse` and that the grpc / futures plumbing arranges to send it over the wire

Yes, although the high-level (i.e. non-protobuf) types on the querier and ingester side are different because they serve different needs. The ingester is mostly concerned about it's own nested structure and how to stream data from it (hence the nested streams) and the querier just wants to collect everything eagerly (since we need the chunks before handing them into the state reconciler and from there into the query engine).

However: The protobuf types and the flattened structure is obviously shared, so the query tests can use the same flattening routine that the gRPC layer uses. The query tests just do the Rust-level type conversion (a rather trivial step).

I'll add some tests for the flattening and the gRPC foo in a follow-up.

**Pull request merged**